File: 40_dynamic_programming/knapsack.py
def knapsack(cap: int, weights: list[int], values: list[int])->int:
    n = len(weights)
    dp = [[0] * (cap + 1) for _ in range(n)]

    for c in range(weights[0], cap+1):
        dp[0][c] = values[0]
    
    for i in range(1, n):
        for j in range(cap+1):
            if j - weights[i] >= 0:
                dp[i][j] = max(dp[i-1][j-weights[i]] + values[i], dp[i-1][j])
            else:
                dp[i][j] = dp[i-1][j]
    
    print(dp)
    return dp[-1][-1]

if __name__ == "__main__":
    cap = 7
    weights = [5, 3, 4, 1]
    values = [70, 50, 40, 10]
    print(knapsack(cap, weights, values))

File: 40_dynamic_programming/knapsack.cpp
#include <algorithm>
#include <iostream>
#include <vector>

int knapsack(int cap, const std::vector<int>& weights, const std::vector<int>& values) {
    int n = static_cast<int>(weights.size());
    std::vector<std::vector<int>> dp(n, std::vector<int>(cap + 1, 0));

    for (int c = weights[0]; c <= cap; c++) {
        dp[0][c] = values[0];
    }

    for (int i = 1; i < n; i++) {
        for (int j = 0; j <= cap; j++) {
            if (j >= weights[i]) {
                dp[i][j] = std::max(dp[i-1][j-weights[i]]+values[i], dp[i-1][j]);
            } else {
                dp[i][j] = dp[i-1][j];
            }
        }
    }

    return dp[n-1][cap];
}

int main() {
    int cap = 7;
    std::vector<int> weights = {5, 3, 4, 1};
    std::vector<int> values = {70, 50, 40, 10};

    std::cout << knapsack(cap, weights, values) << std::endl;

    return 0;
}